source/utilities.tex

```diff
@@ -6463,6 +6464,2036 @@
+      constexpr explicit unexpected(Err&&);
+
+    constexpr unexpected& operator=(const unexpected&) = default;
+    constexpr unexpected& operator=(unexpected&&) = default;
+
+    constexpr const E& value() const& noexcept;
+    constexpr E& value() & noexcept;
+    constexpr const E&& value() const&& noexcept;
+    constexpr E&& value() && noexcept;
+
+    constexpr void swap(unexpected& other) noexcept(@\seebelow@);
+
+    template<class E2>
+      friend constexpr bool operator==(const unexpected&, const unexpected<E2>&);
+
+    friend constexpr void swap(unexpected& x, unexpected& y) noexcept(noexcept(x.swap(y)));
+
+  private:
+    E @\exposid{val}@;               // \expos
+  };
+
+  template<class E> unexpected(E) -> unexpected<E>;
+}
+\end{codeblock}
+
+\pnum
+A program that instantiates the definition of \tcode{unexpected} for
+a non-object type,
+an array type,
+a specialization of \tcode{unexpected}, or
+a cv-qualified type
+is ill-formed.
+
+\rSec4[expected.un.ctor]{Constructors}
+
+\indexlibraryctor{unexpected}%
+\begin{itemdecl}
+template<class Err = E>
+  constexpr explicit unexpected(Err&& e);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\begin{itemize}
+\item
+\tcode{is_same_v<remove_cvref_t<Err>, unexpected>} is \tcode{false}; and
+\item
+\tcode{is_same_v<remove_cvref_t<Err>, in_place_t>} is \tcode{false}; and
+\item
+\tcode{is_constructible_v<E, Err>} is \tcode{true}.
+\end{itemize}
+
+\pnum
+\effects
+Direct-non-list-initializes \exposid{val} with \tcode{std::forward<Err>(e)}.
+
+\pnum
+\throws
+Any exception thrown by the initialization of \exposid{val}.
+\end{itemdescr}
+
+\indexlibraryctor{unexpected}%
+\begin{itemdecl}
+template<class... Args>
+  constexpr explicit unexpected(in_place_t, Args&&...);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{is_constructible_v<E, Args...>} is \tcode{true}.
+
+\pnum
+\effects
+Direct-non-list-initializes
+\exposid{val} with \tcode{std::forward<Args>(args)...}.
+
+\pnum
+\throws
+Any exception thrown by the initialization of \exposid{val}.
+\end{itemdescr}
+
+\indexlibraryctor{unexpected}%
+\begin{itemdecl}
+template<class U, class... Args>
+  constexpr explicit unexpected(in_place_t, initializer_list<U>, Args&&...);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{is_constructible_v<E, initializer_list<U>\&, Args...>} is \tcode{true}.
+
+\pnum
+\effects
+Direct-non-list-initializes
+\exposid{val} with \tcode{std::forward<Args>(args)...}.
+
+\throws
+Any exception thrown by the initialization of \exposid{val}.
+\end{itemdescr}
+
+\rSec4[expected.un.observe]{Observers}
+
+\indexlibrarymember{value}{unexpected}%
+\begin{itemdecl}
+constexpr const E& value() const& noexcept;
+constexpr E& value() & noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\exposid{val}.
+\end{itemdescr}
+
+\indexlibrarymember{value}{unexpected}%
+\begin{itemdecl}
+constexpr E&& value() && noexcept;
+constexpr const E&& value() const&& noexcept;
```

Comment:
```suggestion
constexpr const E&& value() const&& noexcept;
constexpr E&& value() && noexcept;
```
Should we keep this consistent with the general synopsis? (Also many other specifications are inconsistent too, like `expected<T, E>::operator*()`, `bad_excepted_access<E>::error()` etc.. If we fix this, we may as well fix the others too...)

Another way is to fix the general synopsis to be consistent with `const&-&-&&-const&&` order, with the additional benefit of be consistent with `optional<T>::operator*()` and `optional<T>::value()` (these two are the only case in stdlib where all 4 member function qualifiers overload appears currently). However I don't feel this order is nature than, say `const&-&-const&&-&&`. (Maybe we should set a new standard for a natural overload order? i.e. fix `optional<T>` to use `const&-&-const&&-&&` order too?)

Reply:
I thought we usually have a non-const overload first, then const. That's what we do or `begin()` and `end()` in containers and strings, but then we have `operator[]` and `front()` and `back()` in `basic_string` where the `const` overload is first.

I agree we should probably be globally consistent (and IMHO the most natural order is either `&` `const&` `&&` `const&&` or `&` `&&` `const&` `const&&`) but we are already badly inconsistent.

Making this member consistent with the class synopsis makes sense. I don't think we need to rearrange the synopsis for consistency given that we don't have any real consistency.

Reply:
Let's just float the idea: is there any interest inside WG21 editor team for an editorial PR that makes every member function (overloaded only by qualifiers) in stdlib to become a consistent order (presumably `&-const&-&&-const&&`, since the tradition is to specify first two in a group and next two (often need additional `std::move`) in other), and thus we can commit to future consistency in new proposals?

If there is interest I may be able to commit time to make a PR when I have spare time.

Reply:
Can we switch the entire thing to explicit object parameters and avoid the question entirely?
@jwakely , is that an ABI break of concern for the templates we talk about here?

Reply:
I think if the calling convention of
```cpp
struct S {
    void fun(int) const &;
}
```
and
```
struct S {
    static void fun(const S&, int);
```
is completely the same (which it seems to be, testing on godbolt), then there shouldn't be an ABI break (except that the mangled name is different, and this can be solved by providing aliases).

However a possible scenario is that if we provide it as
```cpp
template<typename E>
struct unexpected {
    constexpr decltype(auto) value(this auto&& self) noexcept {return std::forward<decltype(self)>(self).unex;}
}
```
then, I think that for types derivated from `unexpected<E>` this can be an ABI break: the calling convention between calling `unexpected<E>::value() const &` (ordinary member) and calling `unexpected<E>::value<const Derived&>(this const Derived&)` (deducing this) can be different, because the former you need a cast from `const Derived&` to `const unexpected<E>&` (which can be not nop, in case of multiple inheritance), and the latter due to template you do not need any cast.

Reply:
There cannot possibly be an ABI break for `std::unexpected` because it doesn't exist in any implementation yet.

And anyway, if the mangled name is different (and it is) then it doesn't matter whether the calling convention is the same or not. It's a different symbol and existing callers of the old symbol (that already have an instantiation of the old definition) cannot be affected by the addition of a new symbol. The only potential problem is if users are relying on explicit instantiations of `std::optional`, but whether that is supported or not is unspecified and QoI.

But that's all off-topic for an editorial PR.

Reply:
Ah, yeah, sorry I misunderstand the ABI break scenario… Then that’s great and so think we can change optional/expected to explicit this without any breakage.

Anyways, this needs a new proposal to change and is outside the scope. I think for this PR the only thing we need to done is to change every specification order to match the general synopsis…